# Segmenting and Clustering Neighborhoods in Toronto Part 1 to Part 3

in addition to tools contained in libraries such as numpy, panda matplotlib and sklearn, the following tools are required and need to be installed first

In [1]:
!pip install geopy # for importing geographic locations
!pip install folium==0.5.0 #useful for creating maps 

!pip install BeautifulSoup4 # for scraping web based files

print("Packages installed.")

     |████████████████████████████████| 79 kB 7.8 MB/s  eta 0:00:01
  Created wheel for folium: filename=folium-0.5.0-py3-none-any.whl size=76240 sha256=65638eeabe77e75d25cacacd93ff6f641e9da6aad23149e80240791d2364914c
  Stored in directory: /tmp/wsuser/.cache/pip/wheels/b2/2f/2c/109e446b990d663ea5ce9b078b5e7c1a9c45cca91f377080f8
Successfully built folium
Packages installed.


lets import all the libraries and dependencies that we are going to use with the following lines of code

In [2]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
from geopy.geocoders import Nominatim # import geocoder
import requests 
from bs4 import BeautifulSoup 

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans
import folium # map rendering library

print("Libraries imported.")

Libraries imported.


# Part 1 

scrap the wikipedia file that contain data of the Toronto Neighborhood

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M" # url containg data of Toronto neighborhoods
r = requests.get(url) 
   
soup = BeautifulSoup(r.content, "html5lib") 
table = soup.find("div", attrs = {"id":"container"}) 
print("Scrapping Completed")

Scrapping Completed


refine the page scrapped using the following lines of code and define a dataframe so that it is usable by panda

In [5]:
postalCodes = [];
boroughs= [];
neighborhoods = [];
columnNum = 1;
passVal = False

for row in soup.find_all('tr'):
    for cell in row:
        if cell.string and cell.string[0].isalpha() and len(cell.string) > 2:
            passVal = False
            if columnNum == 1:
                if passVal == False and cell.string[1].isdigit():
                    postalCodes.append(cell.string);   
                    columnNum = 2
                else:
                    continue
            elif columnNum == 2 :
                if cell.string == 'Not assigned':
                    passVal = True
                    del postalCodes[-1]
                    columnNum = 1
                    continue
                else:
                    boroughs.append(cell.string);      
                    columnNum = 3
            elif columnNum == 3 :
                if cell.string == 'Not assigned':
                    neighborhoods.append(boroughs[-1])
                else:
                    neighborhoods.append(cell.string); 
                columnNum = 1
                
print("Data Collected.")

Data Collected.


Define the dataframe's columns just created and name the resulting file neighbors

In [6]:
column_names = ['PostalCode', 'Borough', 'Neighborhood'] 
neighbors = pd.DataFrame(columns=column_names)
neighbors

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

now lets populate the columns that are defined with the data collected and the columns just created


In [7]:
for data in range(0, len(postalCodes)-1):
    code = postalCodes[data]
    borough = boroughs[data]
    neighborhood_name = neighborhoods[data]

    neighbors = neighbors.append({ 'PostalCode': code,
                                   'Borough': borough,
                                   'Neighborhood': neighborhood_name}, ignore_index=True)
    
neighbors

PostalCode             Borough  \
0        M1A\n      Not assigned\n   
1        M2A\n      Not assigned\n   
2        M3A\n        North York\n   
3        M4A\n        North York\n   
4        M5A\n  Downtown Toronto\n   
..         ...                 ...   
174      M4Z\n      Not assigned\n   
175      M5Z\n      Not assigned\n   
176      M6Z\n      Not assigned\n   
177      M7Z\n      Not assigned\n   
178      M8Z\n         Etobicoke\n   

                                          Neighborhood  
0                                       Not assigned\n  
1                                       Not assigned\n  
2                                          Parkwoods\n  
3                                   Victoria Village\n  
4                          Regent Park, Harbourfront\n  
..                                                 ...  
174                                     Not assigned\n  
175                                     Not assigned\n  
176                                     Not assigned\n  
177                                     Not assigned\n  
178  Mimico NW, The Queensway West, South of Bloor,...  

[179 rows x 3 columns]

As we can see there are 179 rows and three columns. Moreover, it looks mess with some "\n" in all the rows and there are also number of rows duplicated.

In [9]:
#remove the "\n" from all the rows
neighbors["Neighborhood"] = neighbors["Neighborhood"].str.replace("\n","")
neighbors["PostalCode"] = neighbors["PostalCode"].str.replace("\n","")
neighbors["Borough"] = neighbors["Borough"].str.replace("\n","")
#check if the "\n" is removed from the rows
neighbors.head()

PostalCode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

We need few other tasks before we can proceed our analysis. lets drop the rows where Borough is "Not Assigned" and lets combine neighborhoods with same Postal Code. We also need to replace the name of neighborhood which are "Not Assigned" with the names of the Borough they are located in.

In [26]:
# Dropping the rows where Borough is 'Not assigned'
neighbors = neighbors[neighbors.Borough != "Not assigned"]


In [27]:
# Combining the neighbourhoods with same Postalcode
neighbors = neighbors1.groupby(["PostalCode","Borough"], sort=False).agg(', '.join)
neighbors.reset_index(inplace=True)

In [28]:
# Replacing the name of the neighbourhoods which are 'Not assigned' with names of Borough
neighbors["Neighborhood"] = np.where(neighbors["Neighborhood"] == "Not assigned",neighbors["Borough"], neighbors["Neighborhood"])


In [41]:
# To see changes took effect on the file use
neighbors.head(12)


PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

Examine the shape of the resulting file to see how the changes took effect 

In [30]:
neighbors.shape

(103, 3)

We have now a Toronto neighborhood dataframe file with 103 rows and 3 columns. This answers the first part of the question we are dealing with.

# Part 2 

 To display a dataframe file as a map, we need coordinates such as latitude and longitude. For the dataframe file we scapped from the wiki page, we need additional task of finding coordinates. There are number of ways to get coordinates of the Toronto Neighborhood, but we will use the csv file on the url given. 

In [32]:
#import the csv file that contains the longtude and latitude of each of the neighborhoods in Toronto and name the file as tor_neigh_coor
tor_neigh_coor = pd.read_csv('https://cocl.us/Geospatial_data')
# check if the file is correctly imported and is in a valid data frame format
tor_neigh_coor.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [34]:
#Perfect!, we have imported coordinates of the neighborhoods postal code. lets check the shape of the resulting file
tor_neigh_coor.shape

(103, 3)

Perfect!! again!, we have imported coordinate file for the exact number of neighborhood as the Boroghs above. However, for the purpose of this exercise we need to combine the above two files (neighbors and tor_neigh_coor). 

In [40]:
tor_neighbors = pd.merge(neighbors, tor_neigh_coor, how='left', left_on = 'PostalCode', right_on = 'Postal Code')
# remove the "Postal Code" column as it is duplicated 
tor_neighbors.drop("Postal Code", axis=1, inplace=True)
tor_neighbors.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

In [39]:
# to see if we have done correctely lets check the shape of the toronto neighborhood file.
tor_neighbors.shape

(103, 5)

That Answers the second part of the question, and lets proceed with the third part and perform, segmentation and clustering activities

# Part 3

Creat map of Toronto using latitude and longitude values.

In [45]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="ON_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [47]:
Toronto_map = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(tor_neighbors['Latitude'], tor_neighbors['Longitude'], tor_neighbors['Borough'], tor_neighbors['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(Toronto_map)  
    
Toronto_map

Using FourSquare API Credentails, for data display and segmentation/cluster analysis.

In [49]:
CLIENT_ID = "FMY0ATGGQP3LW2D0CRUCGTMTQSSJYDDC4HCUMMGOGLLHXHJ0"
CLIENT_SECRET = "1HO44Y1NZPVCTZMTMCWFET1BDXWBLDSRFNKHFZWHYZEGAOMX"
VERSION = "20180605" # Foursquare API version

print("Your credentails:")
print("CLIENT_ID: " + CLIENT_ID)
print("CLIENT_SECRET:" + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FMY0ATGGQP3LW2D0CRUCGTMTQSSJYDDC4HCUMMGOGLLHXHJ0
CLIENT_SECRET:1HO44Y1NZPVCTZMTMCWFET1BDXWBLDSRFNKHFZWHYZEGAOMX


Explore data and get venues in 500 meters range from the first Entry

In [50]:
neighborhood_latitude = tor_neighbors.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = tor_neighbors.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = tor_neighbors.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


Creat the GET request url using your credentails 

In [51]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url


'https://api.foursquare.com/v2/venues/explore?&client_id=FMY0ATGGQP3LW2D0CRUCGTMTQSSJYDDC4HCUMMGOGLLHXHJ0&client_secret=1HO44Y1NZPVCTZMTMCWFET1BDXWBLDSRFNKHFZWHYZEGAOMX&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [52]:
#store the regues results json file as tor_results
tor_results = requests.get(url).json()
tor_results 


{'meta': {'code': 200, 'requestId': '60066783e88fa63309765a4b'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

Use the following function to extracts the category of the venue in the neighborhood.

In [53]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

see how many responses are there for the first entry

In [54]:
import json
from pandas.io.json import json_normalize

venues = tor_results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:6: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

Let's creat a function to get information of all neighborhood in Toronto.

In [55]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
  # make the GET request
        tor_results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in tor_results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


In [56]:
toronto_venues = getNearbyVenues(names=tor_neighbors['Neighborhood'],
                                   latitudes=tor_neighbors['Latitude'],
                                   longitudes=tor_neighbors['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Check the dimenstions of the dataframe files.

In [57]:
print(toronto_venues.shape)
toronto_venues.head()

(2129, 7)


Neighbourhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3               Portugril       43.725819       -79.312785   
4             Tim Hortons       43.725517       -79.313103   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3  Portuguese Restaurant  
4            Coffee Shop

Let's see how many venues are there for each Neighborhood

In [58]:
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                            
Agincourt                                                            5   
Alderwood, Long Branch                                               6   
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
...                                                                ...   
Willowdale, Willowdale East                                         35   
Willowdale, Willowdale West                                          5   
Woburn                                                               3   
Woodbine Heights                                                     7   
York Mills West                                                      4   

                                                 Neighborhood Longitude  \
Neighbourhood                                                             
Agincourt                                                             5   
Alderwood, Long Branch                                                6   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale East                                          35   
Willowdale, Willowdale West                                           5   
Woburn                                                                3   
Woodbine Heights                                                      7   
York Mills West                                                       4   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            5               5   
Alderwood, Long Branch                               6               6   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Willowdale East                         35              35   
Willowdale, Willowdale West                          5               5   
Woburn                                               3               3   
Woodbine Heights                                     7               7   
York Mills West                                      4               4   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      5   
Alderwood, Long Branch                                         6   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Willowdale East                                   35   
Willowdale, Willowdale West                                    5   
Woburn                                                         3   
Woodbine Heights                                               7   
York Mills West                                                4   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     5  
Alderwood, Long Branch                                        6 

Lets examine how many categories we can find for each Neighborhood?

In [59]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()


Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Turkish Restaurant  \
0                    0             0  ...              0                   0   
1                    0             0  ...              0                   0   
2                    0             0  ...              0                   0   
3                    0             0  ...              0                   0   
4                    0             0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 272 columns]

Again let's examine by grouping the onehot dataframe by Neighborhood.

In [60]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.045455           0.0  ...            0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 272 columns]

Now Lets Get the top ten for each Neighborhood on the dataframe

In [61]:
# Use the following funcion to define most common venues and sort them in increasing order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [62]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink  Latin American Restaurant        Breakfast Spot   
1                   Gym                Coffee Shop        Sandwich Place   
2                  Bank                       Park    Frozen Yogurt Shop   
3                  Bank        Japanese Restaurant    Chinese Restaurant   
4    Italian Restaurant            Thai Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store            Donut Shop                 Diner   
1                   Pub           Yoga Studio   Distribution Center   
2           Bridal Shop        Sandwich Place                 Diner   
3          Dessert Shop                 Diner        Discount Store   
4              Pharmacy     Indian Restaurant                   Pub   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0        Discount Store   Distribution Center                    Dog Run  
1          Dessert Shop    Dim Sum Restaurant                      Diner  
2            Restaurant         Deli / Bodega  Middle Eastern Restaurant  
3   Distribution Center               Dog Run           Doner Restaurant  
4      Sushi Restaurant          Liquor Store                 Restaurant

Now lets perform clusters of neighborhood, according to their most common characteristics

In [63]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

Merge the created dataframe with the top ten and the resulting cluster for each neighborhood.

In [66]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = tor_neighbors

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighborhood')

toronto_merged.head() 

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0  43.753259 -79.329656             3.0                   Park   
1  43.725882 -79.315572             1.0  Portuguese Restaurant   
2  43.654260 -79.360636             1.0            Coffee Shop   
3  43.718518 -79.464763             1.0         Clothing Store   
4  43.662301 -79.389494             1.0            Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio        Doner Restaurant   
1           Coffee Shop           Pizza Place            Hockey Arena   
2                  Park                Bakery                     Pub   
3     Accessories Store              Boutique  Furniture / Home Store   
4      Sushi Restaurant           Yoga Studio     Japanese Restaurant   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0    Dim Sum Restaurant                  Diner        Discount Store   
1           Yoga Studio           Dessert Shop    Dim Sum Restaurant   
2        Breakfast Spot                   Café               Theater   
3           Event Space  Vietnamese Restaurant           Coffee Shop   
4                   Bar     Mexican Restaurant                  Café   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Distribution Center               Dog Run             Donut Shop  
1                 Diner        Discount Store    Distribution Center  
2            Beer Store            Shoe Store                Brewery  
3   Arts & Crafts Store               Dog Run                  Diner  
4    Italian Restaurant              Beer Bar                  Diner

In [67]:
#we can see two of the rows return with null cluster labels
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

PostalCode      Borough              Neighborhood   Latitude  Longitude  \
45        M2L   North York  York Mills, Silver Hills  43.757490 -79.374714   
95        M1X  Scarborough               Upper Rouge  43.836125 -79.205636   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45             NaN                   NaN                   NaN   
95             NaN                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
45                   NaN                    NaN  
95                   NaN                    NaN

We almost done with the exercise but we need to plot the clusters for better understanding of our analysis.

In [69]:
tor_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

toronto_merged= toronto_merged.dropna(subset=['Cluster Labels'])

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(tor_clusters)
       
tor_clusters

Finally let's examine each of the the 5 clusters

### Cluster 1 

In [71]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6   Scarborough             0.0  Fast Food Restaurant            Print Shop   
11    Etobicoke             0.0                Bakery            Print Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6            Yoga Studio      Doner Restaurant    Dim Sum Restaurant   
11           Yoga Studio      Department Store    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6                  Diner        Discount Store   Distribution Center   
11                 Diner        Discount Store   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
6                Dog Run             Donut Shop  
11               Dog Run       Doner Restaurant

### Cluster 2

In [73]:
# the first five rows are shown but there are 87 rows in this cluster
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head()

Borough  Cluster Labels  1st Most Common Venue  \
1        North York             1.0  Portuguese Restaurant   
2  Downtown Toronto             1.0            Coffee Shop   
3        North York             1.0         Clothing Store   
4  Downtown Toronto             1.0            Coffee Shop   
7        North York             1.0                    Gym   

  2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
1           Coffee Shop           Pizza Place            Hockey Arena   
2                  Park                Bakery                     Pub   
3     Accessories Store              Boutique  Furniture / Home Store   
4      Sushi Restaurant           Yoga Studio     Japanese Restaurant   
7            Beer Store   Japanese Restaurant              Restaurant   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
1           Yoga Studio           Dessert Shop    Dim Sum Restaurant   
2        Breakfast Spot                   Café               Theater   
3           Event Space  Vietnamese Restaurant           Coffee Shop   
4                   Bar     Mexican Restaurant                  Café   
7           Coffee Shop         Discount Store    Italian Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1                 Diner        Discount Store    Distribution Center  
2            Beer Store            Shoe Store                Brewery  
3   Arts & Crafts Store               Dog Run                  Diner  
4    Italian Restaurant              Beer Bar                  Diner  
7                  Café    Dim Sum Restaurant   Caribbean Restaurant

### Cluster 3

In [74]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5  Etobicoke             2.0           Pizza Place           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5            Donut Shop    Dim Sum Restaurant                 Diner   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5        Discount Store   Distribution Center               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
5      Doner Restaurant              Drugstore

### Cluster 4 

In [75]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             3.0                  Park   
21              York             3.0                  Park   
35         East York             3.0                  Park   
52        North York             3.0                  Park   
64              York             3.0                  Park   
66        North York             3.0                  Park   
85       Scarborough             3.0                  Park   
91  Downtown Toronto             3.0                  Park   
98         Etobicoke             3.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop           Yoga Studio      Doner Restaurant   
21         Women's Store                  Pool           Escape Room   
35           Coffee Shop     Convenience Store          Intersection   
52             Piano Bar      Doner Restaurant          Dessert Shop   
64           Yoga Studio            Donut Shop    Dim Sum Restaurant   
66     Convenience Store     Electronics Store            Donut Shop   
85          Intersection            Playground               Dog Run   
91                 Trail            Playground          Dessert Shop   
98                 River           Escape Room     Electronics Store   

          5th Most Common Venue        6th Most Common Venue  \
0            Dim Sum Restaurant                        Diner   
21            Electronics Store  Eastern European Restaurant   
35            Electronics Store                  Escape Room   
52           Dim Sum Restaurant                        Diner   
64                        Diner               Discount Store   
66           Dim Sum Restaurant                        Diner   
85                 Dessert Shop           Dim Sum Restaurant   
91           Dim Sum Restaurant                        Diner   
98  Eastern European Restaurant          Dumpling Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                Discount Store   Distribution Center               Dog Run   
21          Dumpling Restaurant             Drugstore            Donut Shop   
35  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
52               Discount Store   Distribution Center               Dog Run   
64          Distribution Center               Dog Run      Doner Restaurant   
66               Discount Store   Distribution Center               Dog Run   
85                        Diner        Discount Store   Distribution Center   
91               Discount Store   Distribution Center               Dog Run   
98                    Drugstore            Donut Shop         Deli / Bodega   

   10th Most Common Venue  
0              Donut Shop  
21          Deli / Bodega  
35             Donut Shop  
52             Donut Shop  
64              Drugstore  
66       Doner Restaurant  
85       Doner Restaurant  
91       Doner Restaurant  
98                Dog Run

### Cluster 5 

In [76]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
57   North York             4.0  Paper / Office Supplies Store   
101   Etobicoke             4.0                 Baseball Field   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57         Baseball Field           Yoga Studio            Donut Shop   
101           Yoga Studio      Doner Restaurant    Dim Sum Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
57                  Diner        Discount Store   Distribution Center   
101                 Diner        Discount Store   Distribution Center   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
57                Dog Run      Doner Restaurant              Drugstore  
101               Dog Run            Donut Shop     Falafel Restaurant

End of Assignment